balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,629 @@
+			// This should not fail as we should have only one subconn after
+			// entering READY. The subconn should be present in the addressList.
+			b.logger.Errorf("Address %q not found address list in  %v", sd.addr, b.addressList.addresses)
+			return
+		}
+		b.state = connectivity.Ready
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If we are transitioning from READY to IDLE, reset index and re-connect when
+	// prompted.
+	if b.state == connectivity.Ready && state.ConnectivityState == connectivity.Idle {
+		// Once a transport fails, the balancer enters IDLE and starts from
+		// the first address when the picker is used.
+		b.state = connectivity.Idle
+		b.addressList.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch state.ConnectivityState {
+		case connectivity.Connecting:
+			// The balancer can be in either IDLE, CONNECTING or TRANSIENT_FAILURE.
+			// If it's in TRANSIENT_FAILURE, stay in TRANSIENT_FAILURE until
+			// it's READY. See A62.
+			// If the balancer is already in CONNECTING, no update is needed.
+			if b.state == connectivity.Idle {
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			sd.lastErr = state.ConnectionError
+			// Since we're re-using common subconns while handling resolver updates,
+			// we could receive an out of turn TRANSIENT_FAILURE from a pass
+			// over the previous address list. We ignore such updates.
+
+			if curAddr := b.addressList.currentAddress(); !equalAddressIgnoringBalAttributes(&curAddr, &sd.addr) {
+				return
+			}
+			if b.addressList.increment() {
+				b.requestConnection()
+				return
+			}
+			// End of the first pass.
+			b.endFirstPass(state.ConnectionError)
+		case connectivity.Idle:
+			// A subconn can transition from CONNECTING directly to IDLE when
+			// a transport is successfully created, but the connection fails before
+			// the subconn can send the notification for READY. We treat this
+			// as a successful connection and transition to IDLE.
+			if curAddr := b.addressList.currentAddress(); !equalAddressIgnoringBalAttributes(&sd.addr, &curAddr) {
+				return
+			}
+			b.state = connectivity.Idle
+			b.addressList.reset()
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.Idle,
+				Picker:            &idlePicker{exitIdle: b.ExitIdle},
+			})
+		}
+		return
+	}
+
+	// We have finished the first pass, keep re-connecting failing subconns.
+	switch state.ConnectivityState {
+	case connectivity.TransientFailure:
+		sd.lastErr = state.ConnectionError
+		b.cc.UpdateState(balancer.State{
```

Comment:
c-core is updating the picker after seeing `|subchannels|` TFs: https://github.com/grpc/grpc/blob/24e341be62d4411fb7c00c48e8ebc8c538544a0d/src/core/load_balancing/pick_first/pick_first.cc#L898-L907

Java is setting the picker after every TF, but requesting re-resolution after `|subchannels|` TFs: https://github.com/grpc/grpc-java/blob/c63e3548835e838ded2f9eaf3be03dcb6b2b53a7/core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java#L278-L294

Reply:
I think what you describe is what we want -- to update the picker to use the new error message.  If there are queued WFR RPCs then it seems fine to have them go through the picker again and pick up the new error, so they are most accurate when the RPC eventually times out.

Is this behavior not covered in the gRFC?

Reply:
It seems it is covered in part:

> We will count the number of connection failures, and when that number reaches the number of subchannels, we will request re-resolution

Reply:
The gRFC mentions re-resolution after seeing `# of subchannels` failures, but it doesn't say that the picker needs to be updated at the same frequency. In Go presently, the subchannel is requesting re-resolution on every TF. I've opened https://github.com/grpc/grpc-go/issues/7534 to disable re-resolution from the subchannel and control it in PF.

@dfawley to clarify, we want to update the picker for every TF after the first pass?

Reply:
> the subchannel is requesting re-resolution on every TF

True, but TF only occurs after all addresses are attempted.

> to clarify, we want to update the picker for every TF after the first pass?

I'm OK with either behavior.  Every TF means the error gets updated to reflect the latest error, while every N prevents us from updating the picker as often, which means not dequeuing all the pending RPCs as much when we know they'd fail anyway.  I don't think we're ever optimizing for the latter case, but it might be worth asking Mark if he had any good reasons for doing that.

Reply:
Asked Mark, waiting for a response.

Reply:
@dfawley @easwars made the changes to update the picker less frequently, similar to c-core. Added a unit test to verify the same.

Reply:
I thought Eric and Doug wanted more frequent picker updates since that surfaces the recent errors in RPC failures. But I'm OK with leaving this as is since you have gone through the pain of adding a test as well. We can change it later if we decide to go that route.